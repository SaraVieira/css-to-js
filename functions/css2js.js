const { transform } = require("../src/transformers/css2js");

const headers = {
  "Access-Control-Allow-Origin": "*",
  "Access-Control-Allow-Headers": "Content-Type",
};

export async function handler({ body }) {
  return {
    headers,
    statusCode: 200,
    body: JSON.stringify(transform(body)),
  };
}
