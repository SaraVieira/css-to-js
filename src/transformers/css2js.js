import { formatObject } from "../formatters";

const htmlTags = [
  "a",
  "abbr",
  "acronym",
  "address",
  "applet",
  "area",
  "article",
  "aside",
  "audio",
  "b",
  "base",
  "basefont",
  "bdi",
  "bdo",
  "bgsound",
  "big",
  "blink",
  "blockquote",
  "body",
  "br",
  "button",
  "canvas",
  "caption",
  "center",
  "cite",
  "code",
  "col",
  "colgroup",
  "content",
  "data",
  "datalist",
  "dd",
  "decorator",
  "del",
  "details",
  "dfn",
  "dir",
  "div",
  "dl",
  "dt",
  "element",
  "em",
  "embed",
  "fieldset",
  "figcaption",
  "figure",
  "font",
  "footer",
  "form",
  "frame",
  "frameset",
  "h1",
  "h2",
  "h3",
  "h4",
  "h5",
  "h6",
  "head",
  "header",
  "hgroup",
  "hr",
  "html",
  "i",
  "iframe",
  "img",
  "input",
  "ins",
  "isindex",
  "kbd",
  "keygen",
  "label",
  "legend",
  "li",
  "link",
  "listing",
  "main",
  "map",
  "mark",
  "marquee",
  "menu",
  "menuitem",
  "meta",
  "meter",
  "nav",
  "nobr",
  "noframes",
  "noscript",
  "object",
  "ol",
  "optgroup",
  "option",
  "output",
  "p",
  "param",
  "plaintext",
  "pre",
  "progress",
  "q",
  "rp",
  "rt",
  "ruby",
  "s",
  "samp",
  "script",
  "section",
  "select",
  "shadow",
  "small",
  "source",
  "spacer",
  "span",
  "strike",
  "strong",
  "style",
  "sub",
  "summary",
  "sup",
  "table",
  "tbody",
  "td",
  "template",
  "textarea",
  "tfoot",
  "th",
  "thead",
  "time",
  "title",
  "tr",
  "track",
  "tt",
  "u",
  "ul",
  "var",
  "video",
  "wbr",
  "xmp",
];

function checker(value) {
  for (let i = 0; i < htmlTags.length; i++) {
    if (value.startsWith(htmlTags[i])) {
      return true;
    }
  }
  return false;
}

export function transform(css) {
  let code = css.trim() || "";
  if (checker(code) && css.includes("{")) {
    code = css.split("{")[1];
  }

  if (code.startsWith("{")) {
    code = css.substr(1);
  }

  if (code.startsWith(".") || code.startsWith("#")) {
    code = css.split("{")[1];
  }
  const newCode = code ? code.trim().split(";") : [];

  const rules = newCode.reduce((acc, curr) => {
    const both = curr.trim().split(":");
    let prop = both[0];
    let value = both[1];

    // Convert prop from CSS to JS
    if (prop) {
      if (prop.startsWith("-")) {
        prop = `"${prop}"`;
      } else if (prop.includes("-")) {
        prop = prop.replace(/-([a-z])/g, (g) => g[1].toUpperCase());
      }
    }

    // Convert value from CSS to JS
    if (value) {
      if (value.includes("px")) {
        value = parseInt(value.split("px")[0]);
      } else if (value.includes('"')) {
        value = `${value.split('"').join("'")}`;
      } else {
        value = `${value.trim()}`;
      }
    }

    return {
      ...acc,
      ...(!!prop && {
        [prop]: typeof value === "string" ? `"${value.trim()}"` : value,
      }),
    };
  }, {});

  const objString = `{
    ${Object.keys(rules)
      .map((property) => `${property}: ${rules[property]},`)
      .join("\n")}
  }`;

  try {
    return formatObject(objString);
  } catch (e) {
    return objString;
  }
}
