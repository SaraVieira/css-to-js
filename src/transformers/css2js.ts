import { parseCss } from "../utils/parser";
import { formatObject } from "../formatters/js";

// convert props
// -webkit-flex becomes "-webkit-flex"
// margin-bottom becomes marginBottom
function convertCssPropToJs(prop: string): string {
  return prop.startsWith("-")
    ? `"${prop}"`
    : prop.replace(/-([a-z])/g, (g) => g[1].toUpperCase());
}

function convertCssValueToJs(value: string): string {
  return value.replace(/"/g, "'").replace(/(\d)px/g, "$1");
}

export function transform(css: string): any {
  const declarations = parseCss(css);

  const propMap = new Map<string, string | number>();

  declarations.forEach((decl) => {
    let prop: string = convertCssPropToJs(decl.prop);
    let value: string | number = convertCssValueToJs(decl.value);
    const numVal = Number(value);

    // Remove duplicate and maintain
    // position of item
    if (propMap.has(prop)) {
      propMap.delete(prop);
    }
    propMap.set(prop, isNaN(numVal) ? `"${value}"` : numVal);
  });

  const objString: string = `{
    ${[...propMap.keys()]
      .map((key) => `${key}: ${propMap.get(key)}, `)
      .join("\n")}
  }`;

  try {
    return formatObject(objString);
  } catch (e) {
    return objString;
  }
}
