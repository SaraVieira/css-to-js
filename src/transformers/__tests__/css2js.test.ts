import { transform } from "../css2js";

describe("css2js", () => {
  test("transforms a rule with a string value", () => {
    const input = `color: red`;
    expect(transform(input)).toMatchInlineSnapshot(`
      "{
        color: \\"red\\",
      }"
    `);
  });

  test("transforms a rule with a string value with single quotes", () => {
    const input = `font-family: "Inter", sans-serif`;
    expect(transform(input)).toMatchInlineSnapshot(`
      "{
        fontFamily: \\"'Inter', sans-serif\\",
      }"
    `);
  });

  test("transforms a rule with a number value", () => {
    const input = `width: 42px`;
    expect(transform(input)).toMatchInlineSnapshot(`
      "{
        width: 42,
      }"
    `);
  });

  test("transforms a rule with an arbitrary expression value", () => {
    const input = `someProp: someExpression`;
    expect(transform(input)).toMatchInlineSnapshot(`
      "{
        someProp: \\"someExpression\\",
      }"
    `);
  });

  test("transforms a simple rule", () => {
    expect(
      transform(`
        display: block;
        font-size: 16px;
      `)
    ).toMatchInlineSnapshot(`
      "{
        display: \\"block\\",
        fontSize: 16,
      }"
    `);
  });

  test("transforms a more complex rule", () => {
    expect(
      transform(`
        display: block;
        font-size: 16px;
        background: #1e2f5d;
        color: #a4cff4;
        font-family: "Inter", sans-serif;
        font-weight: bold; 
    `)
    ).toMatchInlineSnapshot(`
      "{
        display: \\"block\\",
        fontSize: 16,
        background: \\"#1e2f5d\\",
        color: \\"#a4cff4\\",
        fontFamily: \\"'Inter', sans-serif\\",
        fontWeight: \\"bold\\",
      }"
    `);
  });

  test("transforms multiple rules with classes, ids and elements", () => {
    expect(
      transform(`
        background-color: #fff;
        div {
          color: red;
        }
        #content {
          width: 300px;
        }
        .main {
          height: 100px;
        }
    `)
    ).toMatchInlineSnapshot(`
      "{
        backgroundColor: \\"#fff\\",
        color: \\"red\\",
        width: 300,
        height: 100,
      }"
    `);
  });

  test("transforms rules with duplicate properties and maintains position", () => {
    expect(
      transform(`
        display: block;
        color: red;
        font-size: 16px;
        background: #fff;
        color: green;
        display: flex;
    `)
    ).toMatchInlineSnapshot(`
      "{
        fontSize: 16,
        background: \\"#fff\\",
        color: \\"green\\",
        display: \\"flex\\",
      }"
    `);
  });
});
