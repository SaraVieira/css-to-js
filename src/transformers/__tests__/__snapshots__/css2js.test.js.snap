// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`css2js transforms a more complex rule 1`] = `
"{
  display: \\"block\\",
  fontSize: 16,
  background: \\"#1e2f5d\\",
  color: \\"#a4cff4\\",
  fontFamily: \\"'Inter', sans-serif\\",
  fontWeight: \\"bold\\"
}"
`;

exports[`css2js transforms a simple rule 1`] = `
"{
  display: \\"block\\",
  fontSize: 16
}"
`;
